File: perfume_shop_frontend/lib/services/note_service.dart
import 'dart:convert';
import 'package:http/http.dart' as http;
import '../global/global_parameters.dart';
import '../models/note.dart';

class NoteService {
  // -------------------------------
  // GET ALL NOTES
  // -------------------------------
  Future<List<Note>> getAllNotes() async {
    final response = await http.get(Uri.parse(GlobalParameters.notesEndpoint));

    if (response.statusCode == 200) {
      List<dynamic> data = jsonDecode(response.body);
      return data.map((e) => Note.fromJson(e)).toList();
    } else {
      throw Exception('Failed to load notes');
    }
  }

  // -------------------------------
  // GET NOTE BY ID
  // -------------------------------
  Future<Note?> getNoteById(String id) async {
    final response = await http.get(Uri.parse('${GlobalParameters.notesEndpoint}/$id'));

    if (response.statusCode == 200) {
      return Note.fromJson(jsonDecode(response.body));
    } else if (response.statusCode == 404) {
      return null;
    } else {
      throw Exception('Failed to load note by id');
    }
  }

  // -------------------------------
  // GET NOTE BY NAME
  // -------------------------------
  Future<Note?> getNoteByName(String name) async {
    final response = await http.get(
      Uri.parse('${GlobalParameters.notesEndpoint}/search?name=$name'),
    );

    if (response.statusCode == 200) {
      return Note.fromJson(jsonDecode(response.body));
    } else if (response.statusCode == 404) {
      return null;
    } else {
      throw Exception('Failed to load note by name');
    }
  }

  // -------------------------------
  // CREATE NOTE
  // -------------------------------
  Future<Note> createNote(Note note) async {
    final response = await http.post(
      Uri.parse(GlobalParameters.notesEndpoint),
      headers: {'Content-Type': 'application/json'},
      body: jsonEncode(note.toJson()),
    );

    if (response.statusCode == 201 || response.statusCode == 200) {
      return Note.fromJson(jsonDecode(response.body));
    } else {
      throw Exception('Failed to create note');
    }
  }

  // -------------------------------
  // DELETE NOTE
  // -------------------------------
  Future<void> deleteNote(String id) async {
    final response = await http.delete(Uri.parse('${GlobalParameters.notesEndpoint}/$id'));

    if (response.statusCode != 200 && response.statusCode != 204) {
      throw Exception('Failed to delete note');
    }
  }
}

File: perfume-shop-backend/src/main/java/qa/udst/perfume_shop/service/NoteService.java
package qa.udst.perfume_shop.service;

import lombok.RequiredArgsConstructor;
import org.springframework.stereotype.Service;
import qa.udst.perfume_shop.model.Note;
import qa.udst.perfume_shop.repository.NoteRepository;

import java.util.List;
import java.util.Optional;

@Service
@RequiredArgsConstructor
public class NoteService {

    private final NoteRepository noteRepository;

    // Get all notes
    public List<Note> getAllNotes() {
        return noteRepository.findAll();
    }

    // Get note by ID
    public Optional<Note> getNoteById(String id) {
        return noteRepository.findById(id);
    }

    // Get note by name
    public Optional<Note> getNoteByName(String name) {
        return noteRepository.findByName(name);
    }

    // Create a new note
    public Note createNote(Note note) {
        return noteRepository.save(note);
    }

    // Delete note
    public void deleteNote(String id) {
        noteRepository.deleteById(id);
    }
}
